<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.5"/>
<title>LIMIX: limix.modules.data_deprecated.QTLData Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td style="padding-left: 0.5em;">
   <div id="projectname">LIMIX
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.5 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="pages.html"><span>Related&#160;Pages</span></a></li>
      <li><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li class="current"><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="annotated.html"><span>Class&#160;List</span></a></li>
      <li><a href="classes.html"><span>Class&#160;Index</span></a></li>
      <li><a href="hierarchy.html"><span>Class&#160;Hierarchy</span></a></li>
      <li><a href="functions.html"><span>Class&#160;Members</span></a></li>
    </ul>
  </div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Namespaces</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(5)"><span class="SelectionMark">&#160;</span>Variables</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(6)"><span class="SelectionMark">&#160;</span>Typedefs</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(7)"><span class="SelectionMark">&#160;</span>Enumerations</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(8)"><span class="SelectionMark">&#160;</span>Enumerator</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(9)"><span class="SelectionMark">&#160;</span>Friends</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(10)"><span class="SelectionMark">&#160;</span>Macros</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(11)"><span class="SelectionMark">&#160;</span>Pages</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><a class="el" href="namespacelimix.html">limix</a></li><li class="navelem"><a class="el" href="namespacelimix_1_1modules.html">modules</a></li><li class="navelem"><a class="el" href="namespacelimix_1_1modules_1_1data__deprecated.html">data_deprecated</a></li><li class="navelem"><a class="el" href="classlimix_1_1modules_1_1data__deprecated_1_1_q_t_l_data.html">QTLData</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="#pub-attribs">Public Attributes</a> &#124;
<a href="classlimix_1_1modules_1_1data__deprecated_1_1_q_t_l_data-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">limix.modules.data_deprecated.QTLData Class Reference</div>  </div>
</div><!--header-->
<div class="contents">
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:a039aa0aa8abdc415a15afdce6ccddb65"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__deprecated_1_1_q_t_l_data.html#a039aa0aa8abdc415a15afdce6ccddb65">__init__</a></td></tr>
<tr class="separator:a039aa0aa8abdc415a15afdce6ccddb65"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a688d4aee610ebe7485393c0f45cf61b2"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__deprecated_1_1_q_t_l_data.html#a688d4aee610ebe7485393c0f45cf61b2">load</a></td></tr>
<tr class="memdesc:a688d4aee610ebe7485393c0f45cf61b2"><td class="mdescLeft">&#160;</td><td class="mdescRight">load data file  <a href="#a688d4aee610ebe7485393c0f45cf61b2">More...</a><br/></td></tr>
<tr class="separator:a688d4aee610ebe7485393c0f45cf61b2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3ba4cb55a19b426872f53b241fc5e78a"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__deprecated_1_1_q_t_l_data.html#a3ba4cb55a19b426872f53b241fc5e78a">getGenoIndex</a></td></tr>
<tr class="memdesc:a3ba4cb55a19b426872f53b241fc5e78a"><td class="mdescLeft">&#160;</td><td class="mdescRight">computes 0-based genotype index from position of cumulative position.  <a href="#a3ba4cb55a19b426872f53b241fc5e78a">More...</a><br/></td></tr>
<tr class="separator:a3ba4cb55a19b426872f53b241fc5e78a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a79cf7dee17e61f661c881393c023ba0c"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__deprecated_1_1_q_t_l_data.html#a79cf7dee17e61f661c881393c023ba0c">getGenotypes</a></td></tr>
<tr class="memdesc:a79cf7dee17e61f661c881393c023ba0c"><td class="mdescLeft">&#160;</td><td class="mdescRight">load genotypes.  <a href="#a79cf7dee17e61f661c881393c023ba0c">More...</a><br/></td></tr>
<tr class="separator:a79cf7dee17e61f661c881393c023ba0c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8227d5f1974e1dd2b4ee7465401e35cb"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__deprecated_1_1_q_t_l_data.html#a8227d5f1974e1dd2b4ee7465401e35cb">getCovariance</a></td></tr>
<tr class="memdesc:a8227d5f1974e1dd2b4ee7465401e35cb"><td class="mdescLeft">&#160;</td><td class="mdescRight">calculate the empirical genotype covariance in a region  <a href="#a8227d5f1974e1dd2b4ee7465401e35cb">More...</a><br/></td></tr>
<tr class="separator:a8227d5f1974e1dd2b4ee7465401e35cb"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aab50cffa9efdf48eb3646762178fdcdf"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__deprecated_1_1_q_t_l_data.html#aab50cffa9efdf48eb3646762178fdcdf">getGenoID</a></td></tr>
<tr class="memdesc:aab50cffa9efdf48eb3646762178fdcdf"><td class="mdescLeft">&#160;</td><td class="mdescRight">get genotype IDs.  <a href="#aab50cffa9efdf48eb3646762178fdcdf">More...</a><br/></td></tr>
<tr class="separator:aab50cffa9efdf48eb3646762178fdcdf"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8e21604da28601bcffd94ef385c32697"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__deprecated_1_1_q_t_l_data.html#a8e21604da28601bcffd94ef385c32697">getPhenotypes</a></td></tr>
<tr class="memdesc:a8e21604da28601bcffd94ef385c32697"><td class="mdescLeft">&#160;</td><td class="mdescRight">load Phenotypes  <a href="#a8e21604da28601bcffd94ef385c32697">More...</a><br/></td></tr>
<tr class="separator:a8e21604da28601bcffd94ef385c32697"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aaf4021cc575f9eea049167bdab641684"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__deprecated_1_1_q_t_l_data.html#aaf4021cc575f9eea049167bdab641684">getPos</a></td></tr>
<tr class="memdesc:aaf4021cc575f9eea049167bdab641684"><td class="mdescLeft">&#160;</td><td class="mdescRight">get the positions of the genotypes  <a href="#aaf4021cc575f9eea049167bdab641684">More...</a><br/></td></tr>
<tr class="separator:aaf4021cc575f9eea049167bdab641684"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac1c03b22ce8fa7b091b9eaebfbaaf3d1"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__deprecated_1_1_q_t_l_data.html#ac1c03b22ce8fa7b091b9eaebfbaaf3d1">getIcis_geno</a></td></tr>
<tr class="memdesc:ac1c03b22ce8fa7b091b9eaebfbaaf3d1"><td class="mdescLeft">&#160;</td><td class="mdescRight">if eqtl==True it returns a bool vec for cis  <a href="#ac1c03b22ce8fa7b091b9eaebfbaaf3d1">More...</a><br/></td></tr>
<tr class="separator:ac1c03b22ce8fa7b091b9eaebfbaaf3d1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4e4046f699b03bc40caca6b2e20b0fab"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__deprecated_1_1_q_t_l_data.html#a4e4046f699b03bc40caca6b2e20b0fab">subSample</a></td></tr>
<tr class="memdesc:a4e4046f699b03bc40caca6b2e20b0fab"><td class="mdescLeft">&#160;</td><td class="mdescRight">sample a particular set of individuals (Irow) or phenotypes (Icol_pheno) or genotypes (Icol_geno)  <a href="#a4e4046f699b03bc40caca6b2e20b0fab">More...</a><br/></td></tr>
<tr class="separator:a4e4046f699b03bc40caca6b2e20b0fab"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-attribs"></a>
Public Attributes</h2></td></tr>
<tr class="memitem:aeefce07b736a1851e95c4d1f08b0eaff"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__deprecated_1_1_q_t_l_data.html#aeefce07b736a1851e95c4d1f08b0eaff">file_name</a></td></tr>
<tr class="separator:aeefce07b736a1851e95c4d1f08b0eaff"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a26902b93d709c72419865373de024496"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__deprecated_1_1_q_t_l_data.html#a26902b93d709c72419865373de024496">f</a></td></tr>
<tr class="separator:a26902b93d709c72419865373de024496"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a910af081f5c9190d8239beedffd7f5ef"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__deprecated_1_1_q_t_l_data.html#a910af081f5c9190d8239beedffd7f5ef">pheno</a></td></tr>
<tr class="separator:a910af081f5c9190d8239beedffd7f5ef"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aaa0786bb3b8bfe050e9dab428a79e130"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__deprecated_1_1_q_t_l_data.html#aaa0786bb3b8bfe050e9dab428a79e130">geno</a></td></tr>
<tr class="separator:aaa0786bb3b8bfe050e9dab428a79e130"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad30435d156028d038861e02b9522ae65"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__deprecated_1_1_q_t_l_data.html#ad30435d156028d038861e02b9522ae65">genoM</a></td></tr>
<tr class="separator:ad30435d156028d038861e02b9522ae65"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af77b5440750858a495402c056d0a4cbe"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__deprecated_1_1_q_t_l_data.html#af77b5440750858a495402c056d0a4cbe">phenoM</a></td></tr>
<tr class="separator:af77b5440750858a495402c056d0a4cbe"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:afbe9c774d26d0445443337ed3d8fb67c"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__deprecated_1_1_q_t_l_data.html#afbe9c774d26d0445443337ed3d8fb67c">sample_ID</a></td></tr>
<tr class="separator:afbe9c774d26d0445443337ed3d8fb67c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8cf9be08acc0b05adeddff9bb3dea820"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__deprecated_1_1_q_t_l_data.html#a8cf9be08acc0b05adeddff9bb3dea820">genoChrom</a></td></tr>
<tr class="separator:a8cf9be08acc0b05adeddff9bb3dea820"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af3b3e0eb8322fa5588462e852560351a"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__deprecated_1_1_q_t_l_data.html#af3b3e0eb8322fa5588462e852560351a">genoPos</a></td></tr>
<tr class="separator:af3b3e0eb8322fa5588462e852560351a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0ce74f11705635e7b0a8dc6bdbcaf943"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__deprecated_1_1_q_t_l_data.html#a0ce74f11705635e7b0a8dc6bdbcaf943">genoPos_cum</a></td></tr>
<tr class="separator:a0ce74f11705635e7b0a8dc6bdbcaf943"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6f1185b00a736f5a3a1145409baed79c"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__deprecated_1_1_q_t_l_data.html#a6f1185b00a736f5a3a1145409baed79c">phenotype_ID</a></td></tr>
<tr class="separator:a6f1185b00a736f5a3a1145409baed79c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aad810b48a4b930fe22fc21aca050cee4"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__deprecated_1_1_q_t_l_data.html#aad810b48a4b930fe22fc21aca050cee4">eqtl</a></td></tr>
<tr class="separator:aad810b48a4b930fe22fc21aca050cee4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0a966292c2aa92ea4c86a51e9b23f1b5"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__deprecated_1_1_q_t_l_data.html#a0a966292c2aa92ea4c86a51e9b23f1b5">geneID</a></td></tr>
<tr class="separator:a0a966292c2aa92ea4c86a51e9b23f1b5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aadb124221b3525d00f5247f82edfbb41"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__deprecated_1_1_q_t_l_data.html#aadb124221b3525d00f5247f82edfbb41">gene_pos</a></td></tr>
<tr class="separator:aadb124221b3525d00f5247f82edfbb41"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad51a63dce60283ba0786a2fb5ceeb9a9"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__deprecated_1_1_q_t_l_data.html#ad51a63dce60283ba0786a2fb5ceeb9a9">geneIDs</a></td></tr>
<tr class="separator:ad51a63dce60283ba0786a2fb5ceeb9a9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7f276c5b7a49d8b1449a747a60b05220"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__deprecated_1_1_q_t_l_data.html#a7f276c5b7a49d8b1449a747a60b05220">E</a></td></tr>
<tr class="separator:a7f276c5b7a49d8b1449a747a60b05220"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a15f809866a4890b3510e67b7c28ab8a1"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__deprecated_1_1_q_t_l_data.html#a15f809866a4890b3510e67b7c28ab8a1">Es</a></td></tr>
<tr class="separator:a15f809866a4890b3510e67b7c28ab8a1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7451aa67051b0709a10c963bfe588074"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__deprecated_1_1_q_t_l_data.html#a7451aa67051b0709a10c963bfe588074">N</a></td></tr>
<tr class="separator:a7451aa67051b0709a10c963bfe588074"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a56ebec52898bc7d5acf9648d7648140d"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__deprecated_1_1_q_t_l_data.html#a56ebec52898bc7d5acf9648d7648140d">S</a></td></tr>
<tr class="separator:a56ebec52898bc7d5acf9648d7648140d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a282545c6f6877d7c6ab64dd241e5a00c"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__deprecated_1_1_q_t_l_data.html#a282545c6f6877d7c6ab64dd241e5a00c">P</a></td></tr>
<tr class="separator:a282545c6f6877d7c6ab64dd241e5a00c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4d45b1794e8a16f525de168e0a5ab67b"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__deprecated_1_1_q_t_l_data.html#a4d45b1794e8a16f525de168e0a5ab67b">gneoChrom</a></td></tr>
<tr class="separator:a4d45b1794e8a16f525de168e0a5ab67b"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<h2 class="groupheader">Constructor &amp; Destructor Documentation</h2>
<a class="anchor" id="a039aa0aa8abdc415a15afdce6ccddb65"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def limix.modules.data_deprecated.QTLData.__init__ </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>file_name</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<h2 class="groupheader">Member Function Documentation</h2>
<a class="anchor" id="a8227d5f1974e1dd2b4ee7465401e35cb"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def limix.modules.data_deprecated.QTLData.getCovariance </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>normalize</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>i0</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>i1</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos0</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos1</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>chrom</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>center</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>unit</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_cum0</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_cum1</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>blocksize</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>X</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>kw_args</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>calculate the empirical genotype covariance in a region </p>

</div>
</div>
<a class="anchor" id="aab50cffa9efdf48eb3646762178fdcdf"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def limix.modules.data_deprecated.QTLData.getGenoID </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>i0</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>i1</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos0</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos1</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>chrom</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_cum0</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_cum1</em> = <code>None</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>get genotype IDs. </p>
<pre class="fragment">   Optionally the indices for loading subgroups the genotype IDs for all people
   can be given in one out of three ways:
   - 0-based indexing (i0-i1)
   - position (pos0-pos1 on chrom)
   - cumulative position (pos_cum0-pos_cum1)
   If all these are None (default), then all genotypes are returned
</pre><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">i0</td><td>genotype index based selection (start index) </td></tr>
    <tr><td class="paramname">i1</td><td>genotype index based selection (stop index) </td></tr>
    <tr><td class="paramname">pos0</td><td>position based selection (start position) </td></tr>
    <tr><td class="paramname">pos1</td><td>position based selection (stop position) </td></tr>
    <tr><td class="paramname">chrom</td><td>position based selection (chromosome) </td></tr>
    <tr><td class="paramname">pos_cum0</td><td>cumulative position based selection (start position) </td></tr>
    <tr><td class="paramname">pos_cum1</td><td>cumulative position based selection (stop position)</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd></dd>
<dd>
ID scipy.array of genotype IDs (e.g. rs IDs) </dd></dl>

</div>
</div>
<a class="anchor" id="a3ba4cb55a19b426872f53b241fc5e78a"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def limix.modules.data_deprecated.QTLData.getGenoIndex </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos0</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos1</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>chrom</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_cum0</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_cum1</em> = <code>None</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>computes 0-based genotype index from position of cumulative position. </p>
<pre class="fragment">   Positions can be given in one out of two ways:
   - position (pos0-pos1 on chrom)
   - cumulative position (pos_cum0-pos_cum1)
   If all these are None (default), then all genotypes are returned
</pre><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">pos0</td><td>position based selection (start position) </td></tr>
    <tr><td class="paramname">pos1</td><td>position based selection (stop position) </td></tr>
    <tr><td class="paramname">chrom</td><td>position based selection (chromosome) </td></tr>
    <tr><td class="paramname">pos_cum0</td><td>cumulative position based selection (start position) </td></tr>
    <tr><td class="paramname">pos_cum1</td><td>cumulative position based selection (stop position)</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd></dd>
<dd>
i0 genotype index based selection (start index) </dd>
<dd>
i1 genotype index based selection (stop index) </dd></dl>

</div>
</div>
<a class="anchor" id="a79cf7dee17e61f661c881393c023ba0c"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def limix.modules.data_deprecated.QTLData.getGenotypes </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>i0</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>i1</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos0</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos1</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>chrom</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>center</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>unit</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_cum0</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_cum1</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>impute_missing</em> = <code>True</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>load genotypes. </p>
<pre class="fragment">   Optionally the indices for loading subgroups the genotypes for all people
   can be given in one out of three ways:
   - 0-based indexing (i0-i1)
   - position (pos0-pos1 on chrom)
   - cumulative position (pos_cum0-pos_cum1)
   If all these are None (default), then all genotypes are returned
</pre><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">i0</td><td>genotype index based selection (start index) </td></tr>
    <tr><td class="paramname">i1</td><td>genotype index based selection (stop index) </td></tr>
    <tr><td class="paramname">pos0</td><td>position based selection (start position) </td></tr>
    <tr><td class="paramname">pos1</td><td>position based selection (stop position) </td></tr>
    <tr><td class="paramname">chrom</td><td>position based selection (chromosome) </td></tr>
    <tr><td class="paramname">pos_cum0</td><td>cumulative position based selection (start position) </td></tr>
    <tr><td class="paramname">pos_cum1</td><td>cumulative position based selection (stop position) </td></tr>
    <tr><td class="paramname">impute_missing</td><td>Boolean indicator variable if missing values should be imputed</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd></dd>
<dd>
X scipy.array of genotype values </dd></dl>

</div>
</div>
<a class="anchor" id="ac1c03b22ce8fa7b091b9eaebfbaaf3d1"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def limix.modules.data_deprecated.QTLData.getIcis_geno </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>geneID</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>cis_window</em> = <code>50E3</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>if eqtl==True it returns a bool vec for cis </p>

</div>
</div>
<a class="anchor" id="a8e21604da28601bcffd94ef385c32697"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def limix.modules.data_deprecated.QTLData.getPhenotypes </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>i0</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>i1</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>phenotype_IDs</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>geneIDs</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>environments</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>center</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>impute</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>intersection</em> = <code>False</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>load Phenotypes </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">i0</td><td>phenotype indices to load (start individual index) </td></tr>
    <tr><td class="paramname">i1</td><td>phenotype indices to load (stop individual index) </td></tr>
    <tr><td class="paramname">phenotype_IDs</td><td>names of phenotypes to load </td></tr>
    <tr><td class="paramname">geneIDs</td><td>names of genes to load </td></tr>
    <tr><td class="paramname">environments</td><td>names of environments to load </td></tr>
    <tr><td class="paramname">impute</td><td>imputation of missing values (default: True) </td></tr>
    <tr><td class="paramname">intersection</td><td>restrict observation to those obseved in all phenotypes? (default: False)</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd></dd>
<dd>
Y phenotype values </dd>
<dd>
Ikeep index of individuals in Y </dd></dl>

</div>
</div>
<a class="anchor" id="aaf4021cc575f9eea049167bdab641684"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def limix.modules.data_deprecated.QTLData.getPos </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>get the positions of the genotypes </p>
<dl class="section return"><dt>Returns</dt><dd></dd>
<dd>
chromosome </dd>
<dd>
position </dd>
<dd>
cumulative_position </dd></dl>

</div>
</div>
<a class="anchor" id="a688d4aee610ebe7485393c0f45cf61b2"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def limix.modules.data_deprecated.QTLData.load </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>cache_genotype</em> = <code>False</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>cache_phenotype</em> = <code>True</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>load data file </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">cache_genotype</td><td>load genotypes fully into memory (default: False) </td></tr>
    <tr><td class="paramname">cache_phenotype</td><td>load phentopyes fully intro memry (default: True) </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a class="anchor" id="a4e4046f699b03bc40caca6b2e20b0fab"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def limix.modules.data_deprecated.QTLData.subSample </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>Irow</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>Icol_geno</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>Icol_pheno</em> = <code>None</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>sample a particular set of individuals (Irow) or phenotypes (Icol_pheno) or genotypes (Icol_geno) </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">Irow</td><td>indices for a set of individuals </td></tr>
    <tr><td class="paramname">Icol_pheno</td><td>indices for a set of phenotypes </td></tr>
    <tr><td class="paramname">Icol_geno</td><td>indices for a set of SNPs</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>QTLdata opject holding the specified subset of the data </dd></dl>

</div>
</div>
<h2 class="groupheader">Member Data Documentation</h2>
<a class="anchor" id="a7f276c5b7a49d8b1449a747a60b05220"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">limix.modules.data_deprecated.QTLData.E</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="aad810b48a4b930fe22fc21aca050cee4"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">limix.modules.data_deprecated.QTLData.eqtl</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a15f809866a4890b3510e67b7c28ab8a1"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">limix.modules.data_deprecated.QTLData.Es</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a26902b93d709c72419865373de024496"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">limix.modules.data_deprecated.QTLData.f</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="aeefce07b736a1851e95c4d1f08b0eaff"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">limix.modules.data_deprecated.QTLData.file_name</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="aadb124221b3525d00f5247f82edfbb41"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">limix.modules.data_deprecated.QTLData.gene_pos</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a0a966292c2aa92ea4c86a51e9b23f1b5"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">limix.modules.data_deprecated.QTLData.geneID</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="ad51a63dce60283ba0786a2fb5ceeb9a9"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">limix.modules.data_deprecated.QTLData.geneIDs</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="aaa0786bb3b8bfe050e9dab428a79e130"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">limix.modules.data_deprecated.QTLData.geno</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a8cf9be08acc0b05adeddff9bb3dea820"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">limix.modules.data_deprecated.QTLData.genoChrom</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="ad30435d156028d038861e02b9522ae65"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">limix.modules.data_deprecated.QTLData.genoM</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="af3b3e0eb8322fa5588462e852560351a"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">limix.modules.data_deprecated.QTLData.genoPos</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a0ce74f11705635e7b0a8dc6bdbcaf943"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">limix.modules.data_deprecated.QTLData.genoPos_cum</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a4d45b1794e8a16f525de168e0a5ab67b"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">limix.modules.data_deprecated.QTLData.gneoChrom</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a7451aa67051b0709a10c963bfe588074"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">limix.modules.data_deprecated.QTLData.N</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a282545c6f6877d7c6ab64dd241e5a00c"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">limix.modules.data_deprecated.QTLData.P</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a910af081f5c9190d8239beedffd7f5ef"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">limix.modules.data_deprecated.QTLData.pheno</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="af77b5440750858a495402c056d0a4cbe"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">limix.modules.data_deprecated.QTLData.phenoM</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a6f1185b00a736f5a3a1145409baed79c"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">limix.modules.data_deprecated.QTLData.phenotype_ID</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a56ebec52898bc7d5acf9648d7648140d"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">limix.modules.data_deprecated.QTLData.S</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="afbe9c774d26d0445443337ed3d8fb67c"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">limix.modules.data_deprecated.QTLData.sample_ID</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<hr/>The documentation for this class was generated from the following file:<ul>
<li>/Users/stegle/research/users/stegle/limix/src/interfaces/python/limix/modules/<a class="el" href="data__deprecated_8py.html">data_deprecated.py</a></li>
</ul>
</div><!-- contents -->
<hr>
</body>
</html>
