<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.5"/>
<title>LIMIX: limix.modules.data_new.QTLData Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td style="padding-left: 0.5em;">
   <div id="projectname">LIMIX
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.5 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="pages.html"><span>Related&#160;Pages</span></a></li>
      <li><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li class="current"><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="annotated.html"><span>Class&#160;List</span></a></li>
      <li><a href="classes.html"><span>Class&#160;Index</span></a></li>
      <li><a href="hierarchy.html"><span>Class&#160;Hierarchy</span></a></li>
      <li><a href="functions.html"><span>Class&#160;Members</span></a></li>
    </ul>
  </div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Namespaces</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(5)"><span class="SelectionMark">&#160;</span>Variables</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(6)"><span class="SelectionMark">&#160;</span>Typedefs</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(7)"><span class="SelectionMark">&#160;</span>Enumerations</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(8)"><span class="SelectionMark">&#160;</span>Enumerator</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(9)"><span class="SelectionMark">&#160;</span>Friends</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(10)"><span class="SelectionMark">&#160;</span>Macros</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(11)"><span class="SelectionMark">&#160;</span>Pages</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><a class="el" href="namespacelimix.html">limix</a></li><li class="navelem"><a class="el" href="namespacelimix_1_1modules.html">modules</a></li><li class="navelem"><a class="el" href="namespacelimix_1_1modules_1_1data__new.html">data_new</a></li><li class="navelem"><a class="el" href="classlimix_1_1modules_1_1data__new_1_1_q_t_l_data.html">QTLData</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="#pub-attribs">Public Attributes</a> &#124;
<a href="classlimix_1_1modules_1_1data__new_1_1_q_t_l_data-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">limix.modules.data_new.QTLData Class Reference</div>  </div>
</div><!--header-->
<div class="contents">

<p>This class is a mask to a the contents of a geno_reader and a pheno_reader or a subset thereof.  
 <a href="classlimix_1_1modules_1_1data__new_1_1_q_t_l_data.html#details">More...</a></p>
<div class="dynheader">
Inheritance diagram for limix.modules.data_new.QTLData:</div>
<div class="dyncontent">
 <div class="center">
  <img src="classlimix_1_1modules_1_1data__new_1_1_q_t_l_data.png" usemap="#limix.modules.data_new.QTLData_map" alt=""/>
  <map id="limix.modules.data_new.QTLData_map" name="limix.modules.data_new.QTLData_map">
</map>
 </div></div>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:a0e0fe83de5f10f775679f1a617bd1401"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__new_1_1_q_t_l_data.html#a0e0fe83de5f10f775679f1a617bd1401">__init__</a></td></tr>
<tr class="separator:a0e0fe83de5f10f775679f1a617bd1401"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae3d855a3cb89e84715ec78da62bc9e29"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__new_1_1_q_t_l_data.html#ae3d855a3cb89e84715ec78da62bc9e29">range_query_geno_local</a></td></tr>
<tr class="memdesc:ae3d855a3cb89e84715ec78da62bc9e29"><td class="mdescLeft">&#160;</td><td class="mdescRight">return an index for a range query on the genotypes  <a href="#ae3d855a3cb89e84715ec78da62bc9e29">More...</a><br/></td></tr>
<tr class="separator:ae3d855a3cb89e84715ec78da62bc9e29"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ada1698de828aec5c369a3cd042b4ce69"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__new_1_1_q_t_l_data.html#ada1698de828aec5c369a3cd042b4ce69">range_query_geno</a></td></tr>
<tr class="memdesc:ada1698de828aec5c369a3cd042b4ce69"><td class="mdescLeft">&#160;</td><td class="mdescRight">return an index for a range query on the genotypes  <a href="#ada1698de828aec5c369a3cd042b4ce69">More...</a><br/></td></tr>
<tr class="separator:ada1698de828aec5c369a3cd042b4ce69"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a709b201ed1ee239406bbe70ded810ba0"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__new_1_1_q_t_l_data.html#a709b201ed1ee239406bbe70ded810ba0">getGenotypes</a></td></tr>
<tr class="memdesc:a709b201ed1ee239406bbe70ded810ba0"><td class="mdescLeft">&#160;</td><td class="mdescRight">return genotypes.  <a href="#a709b201ed1ee239406bbe70ded810ba0">More...</a><br/></td></tr>
<tr class="separator:a709b201ed1ee239406bbe70ded810ba0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac00d4b25d9b335283026f91ee256dfbf"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__new_1_1_q_t_l_data.html#ac00d4b25d9b335283026f91ee256dfbf">getCovariance</a></td></tr>
<tr class="memdesc:ac00d4b25d9b335283026f91ee256dfbf"><td class="mdescLeft">&#160;</td><td class="mdescRight">calculate the empirical genotype covariance in a region  <a href="#ac00d4b25d9b335283026f91ee256dfbf">More...</a><br/></td></tr>
<tr class="separator:ac00d4b25d9b335283026f91ee256dfbf"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab0665a9f64fb6fa59228fd9e522e17b7"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__new_1_1_q_t_l_data.html#ab0665a9f64fb6fa59228fd9e522e17b7">getGenoID</a></td></tr>
<tr class="memdesc:ab0665a9f64fb6fa59228fd9e522e17b7"><td class="mdescLeft">&#160;</td><td class="mdescRight">get genotype IDs.  <a href="#ab0665a9f64fb6fa59228fd9e522e17b7">More...</a><br/></td></tr>
<tr class="separator:ab0665a9f64fb6fa59228fd9e522e17b7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae1aa8386dbef77fa7ac92f0d5a5b500b"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__new_1_1_q_t_l_data.html#ae1aa8386dbef77fa7ac92f0d5a5b500b">getPhenotypes</a></td></tr>
<tr class="memdesc:ae1aa8386dbef77fa7ac92f0d5a5b500b"><td class="mdescLeft">&#160;</td><td class="mdescRight">load Phenotypes  <a href="#ae1aa8386dbef77fa7ac92f0d5a5b500b">More...</a><br/></td></tr>
<tr class="separator:ae1aa8386dbef77fa7ac92f0d5a5b500b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a924630fdf496cdd2785d29f22013adcb"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__new_1_1_q_t_l_data.html#a924630fdf496cdd2785d29f22013adcb">getPos</a></td></tr>
<tr class="memdesc:a924630fdf496cdd2785d29f22013adcb"><td class="mdescLeft">&#160;</td><td class="mdescRight">get the positions of the genotypes  <a href="#a924630fdf496cdd2785d29f22013adcb">More...</a><br/></td></tr>
<tr class="separator:a924630fdf496cdd2785d29f22013adcb"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5e7ae6d9cb94663d7c24835a071513ad"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__new_1_1_q_t_l_data.html#a5e7ae6d9cb94663d7c24835a071513ad">subsample</a></td></tr>
<tr class="memdesc:a5e7ae6d9cb94663d7c24835a071513ad"><td class="mdescLeft">&#160;</td><td class="mdescRight">sample a particular set of individuals (rows) or phenotypes (cols_pheno) or genotypes (cols_geno)  <a href="#a5e7ae6d9cb94663d7c24835a071513ad">More...</a><br/></td></tr>
<tr class="separator:a5e7ae6d9cb94663d7c24835a071513ad"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aeccfe703584ea255341297ef85706a67"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__new_1_1_q_t_l_data.html#aeccfe703584ea255341297ef85706a67">subsample_phenotypes</a></td></tr>
<tr class="memdesc:aeccfe703584ea255341297ef85706a67"><td class="mdescLeft">&#160;</td><td class="mdescRight">load Phenotypes  <a href="#aeccfe703584ea255341297ef85706a67">More...</a><br/></td></tr>
<tr class="separator:aeccfe703584ea255341297ef85706a67"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-attribs"></a>
Public Attributes</h2></td></tr>
<tr class="memitem:a2be6677a4fc02c0dfe5e31f43cb7f069"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__new_1_1_q_t_l_data.html#a2be6677a4fc02c0dfe5e31f43cb7f069">geno_reader</a></td></tr>
<tr class="separator:a2be6677a4fc02c0dfe5e31f43cb7f069"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a42ad86275e949ab9388e73e799418fad"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__new_1_1_q_t_l_data.html#a42ad86275e949ab9388e73e799418fad">pheno_reader</a></td></tr>
<tr class="separator:a42ad86275e949ab9388e73e799418fad"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab69a912fc4ee602701c086ce21cf93b6"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__new_1_1_q_t_l_data.html#ab69a912fc4ee602701c086ce21cf93b6">geno_pos</a></td></tr>
<tr class="separator:ab69a912fc4ee602701c086ce21cf93b6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a91f3c08aceab9d14a9192c5c4adf4491"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__new_1_1_q_t_l_data.html#a91f3c08aceab9d14a9192c5c4adf4491">geno_ID</a></td></tr>
<tr class="separator:a91f3c08aceab9d14a9192c5c4adf4491"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac1af45c981477c0064badbe4279352c5"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__new_1_1_q_t_l_data.html#ac1af45c981477c0064badbe4279352c5">phenotype_ID</a></td></tr>
<tr class="separator:ac1af45c981477c0064badbe4279352c5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:adc093101ae2684a8372ef6e0fdf31409"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__new_1_1_q_t_l_data.html#adc093101ae2684a8372ef6e0fdf31409">geno_snp_idx</a></td></tr>
<tr class="separator:adc093101ae2684a8372ef6e0fdf31409"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1d1a67c2712cbc13319307f8f62cd669"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__new_1_1_q_t_l_data.html#a1d1a67c2712cbc13319307f8f62cd669">sample_idx</a></td></tr>
<tr class="separator:a1d1a67c2712cbc13319307f8f62cd669"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a983b8ea072ab89b23b347e97601a1ff0"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__new_1_1_q_t_l_data.html#a983b8ea072ab89b23b347e97601a1ff0">sample_ID</a></td></tr>
<tr class="separator:a983b8ea072ab89b23b347e97601a1ff0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a84cee901dcf4887c980c45a9b91c72ab"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data__new_1_1_q_t_l_data.html#a84cee901dcf4887c980c45a9b91c72ab">num_snps</a></td></tr>
<tr class="separator:a84cee901dcf4887c980c45a9b91c72ab"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>This class is a mask to a the contents of a geno_reader and a pheno_reader or a subset thereof. </p>
</div><h2 class="groupheader">Constructor &amp; Destructor Documentation</h2>
<a class="anchor" id="a0e0fe83de5f10f775679f1a617bd1401"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def limix.modules.data_new.QTLData.__init__ </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>geno_reader</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pheno_reader</em> = <code>None</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<h2 class="groupheader">Member Function Documentation</h2>
<a class="anchor" id="ac00d4b25d9b335283026f91ee256dfbf"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def limix.modules.data_new.QTLData.getCovariance </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>normalize</em> = <code>False</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>idx_start</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>idx_end</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_start</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_end</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>chrom</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>center</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>unit</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_cum_start</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_cum_end</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>blocksize</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>X</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>kw_args</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>calculate the empirical genotype covariance in a region </p>

</div>
</div>
<a class="anchor" id="ab0665a9f64fb6fa59228fd9e522e17b7"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def limix.modules.data_new.QTLData.getGenoID </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>idx_start</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>idx_end</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_start</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_end</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>chrom</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_cum_start</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_cum_end</em> = <code>None</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>get genotype IDs. </p>
<pre class="fragment">   Optionally the indices for loading subgroups the genotype IDs for all people
   can be given in one out of three ways:
   - 0-based indexing (idx_start-idx_end)
   - position (pos_start-pos_end on chrom)
   - cumulative position (pos_cum_start-pos_cum_end)
   If all these are None (default), then all genotypes are returned
</pre><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">idx_start</td><td>genotype index based selection (start index) </td></tr>
    <tr><td class="paramname">idx_end</td><td>genotype index based selection (end index) </td></tr>
    <tr><td class="paramname">pos_start</td><td>position based selection (start position) </td></tr>
    <tr><td class="paramname">pos_end</td><td>position based selection (end position) </td></tr>
    <tr><td class="paramname">chrom</td><td>position based selection (chromosome) </td></tr>
    <tr><td class="paramname">pos_cum_start</td><td>cumulative position based selection (start position) </td></tr>
    <tr><td class="paramname">pos_cum_end</td><td>cumulative position based selection (end position)</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd></dd>
<dd>
ID scipy.array of genotype IDs (e.g. rs IDs) </dd></dl>

</div>
</div>
<a class="anchor" id="a709b201ed1ee239406bbe70ded810ba0"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def limix.modules.data_new.QTLData.getGenotypes </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>idx_start</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>idx_end</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_start</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_end</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>chrom</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>center</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>unit</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_cum_start</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_cum_end</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>impute_missing</em> = <code>False</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>return genotypes. </p>
<pre class="fragment">   Optionally the indices for loading subgroups the genotypes for all people
   can be given in one out of three ways:
   - 0-based indexing (idx_start-idx_end)
   - position (pos_start-pos_end on chrom)
   - cumulative position (pos_cum_start-pos_cum_end)
   If all these are None (default), then all genotypes are returned
</pre><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">idx_start</td><td>genotype index based selection (start index) </td></tr>
    <tr><td class="paramname">idx_end</td><td>genotype index based selection (end index) </td></tr>
    <tr><td class="paramname">pos_start</td><td>position based selection (start position) </td></tr>
    <tr><td class="paramname">pos_end</td><td>position based selection (end position) </td></tr>
    <tr><td class="paramname">chrom</td><td>position based selection (chromosome) </td></tr>
    <tr><td class="paramname">pos_cum_start</td><td>cumulative position based selection (start position) </td></tr>
    <tr><td class="paramname">pos_cum_end</td><td>cumulative position based selection (end position) </td></tr>
    <tr><td class="paramname">impute_missing</td><td>Boolean indicator variable if missing values should be imputed</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd></dd>
<dd>
X scipy.array of genotype values </dd></dl>

</div>
</div>
<a class="anchor" id="ae1aa8386dbef77fa7ac92f0d5a5b500b"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def limix.modules.data_new.QTLData.getPhenotypes </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>phenotype_IDs</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>center</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>impute</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>intersection</em> = <code>False</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>load Phenotypes </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">idx_start</td><td>phenotype indices to load (start individual index) </td></tr>
    <tr><td class="paramname">idx_end</td><td>phenotype indices to load (end individual index) </td></tr>
    <tr><td class="paramname">phenotype_IDs</td><td>names of phenotypes to load </td></tr>
    <tr><td class="paramname">impute</td><td>imputation of missing values (default: True) </td></tr>
    <tr><td class="paramname">intersection</td><td>restrict observation to those obseved in all phenotypes (true) or at least in one phenotype (false)? (default: False)</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd></dd>
<dd>
phenotypes phenotype values </dd>
<dd>
sample_idx_intersect index of individuals in phenotypes after filtering missing values </dd></dl>

</div>
</div>
<a class="anchor" id="a924630fdf496cdd2785d29f22013adcb"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def limix.modules.data_new.QTLData.getPos </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>idx_start</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>idx_end</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_start</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_end</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>chrom</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_cum_start</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_cum_end</em> = <code>None</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>get the positions of the genotypes </p>
<dl class="section return"><dt>Returns</dt><dd></dd>
<dd>
chromosome </dd>
<dd>
position </dd>
<dd>
cumulative_position </dd></dl>

</div>
</div>
<a class="anchor" id="ada1698de828aec5c369a3cd042b4ce69"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def limix.modules.data_new.QTLData.range_query_geno </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>idx_start</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>idx_end</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>chrom</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_start</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_end</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_cum_start</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_cum_end</em> = <code>None</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>return an index for a range query on the genotypes </p>

</div>
</div>
<a class="anchor" id="ae3d855a3cb89e84715ec78da62bc9e29"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def limix.modules.data_new.QTLData.range_query_geno_local </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>idx_start</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>idx_end</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>chrom</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_start</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_end</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_cum_start</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_cum_end</em> = <code>None</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>return an index for a range query on the genotypes </p>

</div>
</div>
<a class="anchor" id="a5e7ae6d9cb94663d7c24835a071513ad"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def limix.modules.data_new.QTLData.subsample </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>rows</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>cols_pheno</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>cols_geno</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>idx_start</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>idx_end</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_start</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_end</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>chrom</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_cum_start</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_cum_end</em> = <code>None</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>sample a particular set of individuals (rows) or phenotypes (cols_pheno) or genotypes (cols_geno) </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">rows</td><td>indices for a set of individuals </td></tr>
    <tr><td class="paramname">cols_pheno</td><td>indices for a set of phenotypes </td></tr>
    <tr><td class="paramname">cols_geno</td><td>indices for a set of SNPs</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>QTLdata object holding the specified subset of the data </dd></dl>

</div>
</div>
<a class="anchor" id="aeccfe703584ea255341297ef85706a67"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def limix.modules.data_new.QTLData.subsample_phenotypes </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>phenotype_IDs</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>center</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>impute</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>intersection</em> = <code>False</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>load Phenotypes </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">idx_start</td><td>phenotype indices to load (start individual index) </td></tr>
    <tr><td class="paramname">idx_end</td><td>phenotype indices to load (end individual index) </td></tr>
    <tr><td class="paramname">phenotype_IDs</td><td>names of phenotypes to load </td></tr>
    <tr><td class="paramname">impute</td><td>imputation of missing values (default: True) </td></tr>
    <tr><td class="paramname">intersection</td><td>restrict observation to those obseved in all phenotypes (true) or at least in one phenotype (false)? (default: False)</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd></dd>
<dd>
phenotypes phenotype values </dd>
<dd>
sample_idx_intersect index of individuals in phenotypes after filtering missing values </dd></dl>

</div>
</div>
<h2 class="groupheader">Member Data Documentation</h2>
<a class="anchor" id="a91f3c08aceab9d14a9192c5c4adf4491"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">limix.modules.data_new.QTLData.geno_ID</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="ab69a912fc4ee602701c086ce21cf93b6"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">limix.modules.data_new.QTLData.geno_pos</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a2be6677a4fc02c0dfe5e31f43cb7f069"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">limix.modules.data_new.QTLData.geno_reader</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="adc093101ae2684a8372ef6e0fdf31409"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">limix.modules.data_new.QTLData.geno_snp_idx</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a84cee901dcf4887c980c45a9b91c72ab"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">limix.modules.data_new.QTLData.num_snps</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a42ad86275e949ab9388e73e799418fad"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">limix.modules.data_new.QTLData.pheno_reader</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="ac1af45c981477c0064badbe4279352c5"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">limix.modules.data_new.QTLData.phenotype_ID</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a983b8ea072ab89b23b347e97601a1ff0"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">limix.modules.data_new.QTLData.sample_ID</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a1d1a67c2712cbc13319307f8f62cd669"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">limix.modules.data_new.QTLData.sample_idx</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<hr/>The documentation for this class was generated from the following file:<ul>
<li>/Users/stegle/research/users/stegle/limix/src/interfaces/python/limix/modules/<a class="el" href="data__new_8py.html">data_new.py</a></li>
</ul>
</div><!-- contents -->
<hr>
</body>
</html>
