<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.5"/>
<title>LIMIX: limix.modules.data.QTLData Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td style="padding-left: 0.5em;">
   <div id="projectname">LIMIX
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.5 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="pages.html"><span>Related&#160;Pages</span></a></li>
      <li><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li class="current"><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="annotated.html"><span>Class&#160;List</span></a></li>
      <li><a href="classes.html"><span>Class&#160;Index</span></a></li>
      <li><a href="hierarchy.html"><span>Class&#160;Hierarchy</span></a></li>
      <li><a href="functions.html"><span>Class&#160;Members</span></a></li>
    </ul>
  </div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Namespaces</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(5)"><span class="SelectionMark">&#160;</span>Variables</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(6)"><span class="SelectionMark">&#160;</span>Typedefs</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(7)"><span class="SelectionMark">&#160;</span>Enumerations</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(8)"><span class="SelectionMark">&#160;</span>Enumerator</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(9)"><span class="SelectionMark">&#160;</span>Friends</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(10)"><span class="SelectionMark">&#160;</span>Macros</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(11)"><span class="SelectionMark">&#160;</span>Pages</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><a class="el" href="namespacelimix.html">limix</a></li><li class="navelem"><a class="el" href="namespacelimix_1_1modules.html">modules</a></li><li class="navelem"><a class="el" href="namespacelimix_1_1modules_1_1data.html">data</a></li><li class="navelem"><a class="el" href="classlimix_1_1modules_1_1data_1_1_q_t_l_data.html">QTLData</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="#pub-attribs">Public Attributes</a> &#124;
<a href="classlimix_1_1modules_1_1data_1_1_q_t_l_data-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">limix.modules.data.QTLData Class Reference</div>  </div>
</div><!--header-->
<div class="contents">

<p>This class is a mask to a the contents of a geno_reader and a pheno_reader or a subset thereof.  
 <a href="classlimix_1_1modules_1_1data_1_1_q_t_l_data.html#details">More...</a></p>
<div class="dynheader">
Inheritance diagram for limix.modules.data.QTLData:</div>
<div class="dyncontent">
 <div class="center">
  <img src="classlimix_1_1modules_1_1data_1_1_q_t_l_data.png" usemap="#limix.modules.data.QTLData_map" alt=""/>
  <map id="limix.modules.data.QTLData_map" name="limix.modules.data.QTLData_map">
</map>
 </div></div>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:a8f82481be25ffeea630078b206b47f2c"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data_1_1_q_t_l_data.html#a8f82481be25ffeea630078b206b47f2c">__init__</a></td></tr>
<tr class="separator:a8f82481be25ffeea630078b206b47f2c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac5157b8a4ae4a80e933a07c78657cf82"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data_1_1_q_t_l_data.html#ac5157b8a4ae4a80e933a07c78657cf82">range_query_geno_local</a></td></tr>
<tr class="memdesc:ac5157b8a4ae4a80e933a07c78657cf82"><td class="mdescLeft">&#160;</td><td class="mdescRight">return an index for a range query on the genotypes  <a href="#ac5157b8a4ae4a80e933a07c78657cf82">More...</a><br/></td></tr>
<tr class="separator:ac5157b8a4ae4a80e933a07c78657cf82"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abd07aa418ffd48a352b0661748ef63df"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data_1_1_q_t_l_data.html#abd07aa418ffd48a352b0661748ef63df">range_query_geno</a></td></tr>
<tr class="memdesc:abd07aa418ffd48a352b0661748ef63df"><td class="mdescLeft">&#160;</td><td class="mdescRight">return an index for a range query on the genotypes  <a href="#abd07aa418ffd48a352b0661748ef63df">More...</a><br/></td></tr>
<tr class="separator:abd07aa418ffd48a352b0661748ef63df"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:adf7a02599043716cf94bfd09cbc1d9dc"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data_1_1_q_t_l_data.html#adf7a02599043716cf94bfd09cbc1d9dc">getGenotypes</a></td></tr>
<tr class="memdesc:adf7a02599043716cf94bfd09cbc1d9dc"><td class="mdescLeft">&#160;</td><td class="mdescRight">return genotypes.  <a href="#adf7a02599043716cf94bfd09cbc1d9dc">More...</a><br/></td></tr>
<tr class="separator:adf7a02599043716cf94bfd09cbc1d9dc"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a907075432bb8b180479cb33dc7c525f2"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data_1_1_q_t_l_data.html#a907075432bb8b180479cb33dc7c525f2">getCovariance</a></td></tr>
<tr class="memdesc:a907075432bb8b180479cb33dc7c525f2"><td class="mdescLeft">&#160;</td><td class="mdescRight">calculate the empirical genotype covariance in a region  <a href="#a907075432bb8b180479cb33dc7c525f2">More...</a><br/></td></tr>
<tr class="separator:a907075432bb8b180479cb33dc7c525f2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2ef915649cecdeac6d1cb37e8986355d"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data_1_1_q_t_l_data.html#a2ef915649cecdeac6d1cb37e8986355d">getGenoID</a></td></tr>
<tr class="memdesc:a2ef915649cecdeac6d1cb37e8986355d"><td class="mdescLeft">&#160;</td><td class="mdescRight">get genotype IDs.  <a href="#a2ef915649cecdeac6d1cb37e8986355d">More...</a><br/></td></tr>
<tr class="separator:a2ef915649cecdeac6d1cb37e8986355d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9196cd126b754c6f7e83030b0a1b9af8"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data_1_1_q_t_l_data.html#a9196cd126b754c6f7e83030b0a1b9af8">getPhenotypes</a></td></tr>
<tr class="memdesc:a9196cd126b754c6f7e83030b0a1b9af8"><td class="mdescLeft">&#160;</td><td class="mdescRight">load Phenotypes  <a href="#a9196cd126b754c6f7e83030b0a1b9af8">More...</a><br/></td></tr>
<tr class="separator:a9196cd126b754c6f7e83030b0a1b9af8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5251a463471ebf0e1d251fdfc1eec6f7"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data_1_1_q_t_l_data.html#a5251a463471ebf0e1d251fdfc1eec6f7">getPos</a></td></tr>
<tr class="memdesc:a5251a463471ebf0e1d251fdfc1eec6f7"><td class="mdescLeft">&#160;</td><td class="mdescRight">get the positions of the genotypes  <a href="#a5251a463471ebf0e1d251fdfc1eec6f7">More...</a><br/></td></tr>
<tr class="separator:a5251a463471ebf0e1d251fdfc1eec6f7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af193e0fd43b38bf0c4ac40042c431753"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data_1_1_q_t_l_data.html#af193e0fd43b38bf0c4ac40042c431753">subsample</a></td></tr>
<tr class="memdesc:af193e0fd43b38bf0c4ac40042c431753"><td class="mdescLeft">&#160;</td><td class="mdescRight">sample a particular set of individuals (rows) or phenotypes (cols_pheno) or genotypes (cols_geno)  <a href="#af193e0fd43b38bf0c4ac40042c431753">More...</a><br/></td></tr>
<tr class="separator:af193e0fd43b38bf0c4ac40042c431753"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a52831f8f693a82c852ef68bb8b709528"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data_1_1_q_t_l_data.html#a52831f8f693a82c852ef68bb8b709528">subsample_phenotypes</a></td></tr>
<tr class="memdesc:a52831f8f693a82c852ef68bb8b709528"><td class="mdescLeft">&#160;</td><td class="mdescRight">load Phenotypes  <a href="#a52831f8f693a82c852ef68bb8b709528">More...</a><br/></td></tr>
<tr class="separator:a52831f8f693a82c852ef68bb8b709528"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-attribs"></a>
Public Attributes</h2></td></tr>
<tr class="memitem:a9261d716439ec35ab451cf6c6844974b"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data_1_1_q_t_l_data.html#a9261d716439ec35ab451cf6c6844974b">geno_reader</a></td></tr>
<tr class="separator:a9261d716439ec35ab451cf6c6844974b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6244c75260ec6d8462f412eab4602c1a"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data_1_1_q_t_l_data.html#a6244c75260ec6d8462f412eab4602c1a">pheno_reader</a></td></tr>
<tr class="separator:a6244c75260ec6d8462f412eab4602c1a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a89f88a15267238a5891760f65dec36ba"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data_1_1_q_t_l_data.html#a89f88a15267238a5891760f65dec36ba">geno_pos</a></td></tr>
<tr class="separator:a89f88a15267238a5891760f65dec36ba"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a44790730592493cfead5620fa01fd2fe"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data_1_1_q_t_l_data.html#a44790730592493cfead5620fa01fd2fe">geno_ID</a></td></tr>
<tr class="separator:a44790730592493cfead5620fa01fd2fe"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a921a453d6b5484595508867ea848459d"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data_1_1_q_t_l_data.html#a921a453d6b5484595508867ea848459d">phenotype_ID</a></td></tr>
<tr class="separator:a921a453d6b5484595508867ea848459d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8ec02bd88a4802018e4f50c5ae93f0bc"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data_1_1_q_t_l_data.html#a8ec02bd88a4802018e4f50c5ae93f0bc">geno_snp_idx</a></td></tr>
<tr class="separator:a8ec02bd88a4802018e4f50c5ae93f0bc"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6882068a6f98048ae7ba8ee132b4523d"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data_1_1_q_t_l_data.html#a6882068a6f98048ae7ba8ee132b4523d">sample_idx</a></td></tr>
<tr class="separator:a6882068a6f98048ae7ba8ee132b4523d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa205ab51d0684b45bdb25f476afd0256"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data_1_1_q_t_l_data.html#aa205ab51d0684b45bdb25f476afd0256">sample_ID</a></td></tr>
<tr class="separator:aa205ab51d0684b45bdb25f476afd0256"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a699163ea6b952f2d30e08f270a2383da"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classlimix_1_1modules_1_1data_1_1_q_t_l_data.html#a699163ea6b952f2d30e08f270a2383da">num_snps</a></td></tr>
<tr class="separator:a699163ea6b952f2d30e08f270a2383da"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>This class is a mask to a the contents of a geno_reader and a pheno_reader or a subset thereof. </p>
</div><h2 class="groupheader">Constructor &amp; Destructor Documentation</h2>
<a class="anchor" id="a8f82481be25ffeea630078b206b47f2c"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def limix.modules.data.QTLData.__init__ </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>geno_reader</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pheno_reader</em> = <code>None</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<h2 class="groupheader">Member Function Documentation</h2>
<a class="anchor" id="a907075432bb8b180479cb33dc7c525f2"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def limix.modules.data.QTLData.getCovariance </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>normalize</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>idx_start</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>idx_end</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_start</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_end</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>windowsize</em> = <code>0</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>chrom</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>center</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>unit</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>blocksize</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>X</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>kw_args</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>calculate the empirical genotype covariance in a region </p>

</div>
</div>
<a class="anchor" id="a2ef915649cecdeac6d1cb37e8986355d"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def limix.modules.data.QTLData.getGenoID </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>idx_start</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>idx_end</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_start</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_end</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>chrom</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>windowsize</em> = <code>0</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>get genotype IDs. </p>
<pre class="fragment">   Optionally the indices for loading subgroups the genotype IDs for all people
   can be given in one out of three ways:
   - 0-based indexing (idx_start-idx_end)
   - position (pos_start-pos_end on chrom)
   - cumulative position (pos_cum_start-pos_cum_end)
   If all these are None (default), then all genotypes are returned
</pre><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">idx_start</td><td>genotype index based selection (start index) </td></tr>
    <tr><td class="paramname">idx_end</td><td>genotype index based selection (end index) </td></tr>
    <tr><td class="paramname">pos_start</td><td>position based selection (start position) </td></tr>
    <tr><td class="paramname">pos_end</td><td>position based selection (end position) </td></tr>
    <tr><td class="paramname">chrom</td><td>position based selection (chromosome) </td></tr>
    <tr><td class="paramname">pos_cum_start</td><td>cumulative position based selection (start position) </td></tr>
    <tr><td class="paramname">pos_cum_end</td><td>cumulative position based selection (end position)</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd></dd>
<dd>
ID scipy.array of genotype IDs (e.g. rs IDs) </dd></dl>

</div>
</div>
<a class="anchor" id="adf7a02599043716cf94bfd09cbc1d9dc"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def limix.modules.data.QTLData.getGenotypes </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>idx_start</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>idx_end</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_start</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_end</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>windowsize</em> = <code>0</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>chrom</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>center</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>unit</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>impute_missing</em> = <code>False</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>return genotypes. </p>
<pre class="fragment">   Optionally the indices for loading subgroups the genotypes for all people
   can be given in one out of three ways:
   - 0-based indexing (idx_start-idx_end)
   - position (pos_start-pos_end on chrom)
   - cumulative position (pos_cum_start-pos_cum_end)
   If all these are None (default), then all genotypes are returned
</pre><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">idx_start</td><td>genotype index based selection (start index) </td></tr>
    <tr><td class="paramname">idx_end</td><td>genotype index based selection (end index) </td></tr>
    <tr><td class="paramname">pos_start</td><td>position based selection (start position) </td></tr>
    <tr><td class="paramname">pos_end</td><td>position based selection (end position) </td></tr>
    <tr><td class="paramname">chrom</td><td>position based selection (chromosome) </td></tr>
    <tr><td class="paramname">pos_cum_start</td><td>cumulative position based selection (start position) </td></tr>
    <tr><td class="paramname">pos_cum_end</td><td>cumulative position based selection (end position) </td></tr>
    <tr><td class="paramname">impute_missing</td><td>Boolean indicator variable if missing values should be imputed</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd></dd>
<dd>
X scipy.array of genotype values </dd></dl>

</div>
</div>
<a class="anchor" id="a9196cd126b754c6f7e83030b0a1b9af8"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def limix.modules.data.QTLData.getPhenotypes </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>phenotype_IDs</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>phenotype_query</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>center</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>intersection</em> = <code>False</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>load Phenotypes </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">idx_start</td><td>phenotype indices to load (start individual index) </td></tr>
    <tr><td class="paramname">idx_end</td><td>phenotype indices to load (end individual index) </td></tr>
    <tr><td class="paramname">phenotype_IDs</td><td>names of phenotypes to load </td></tr>
    <tr><td class="paramname">impute</td><td>imputation of missing values (default: True) </td></tr>
    <tr><td class="paramname">intersection</td><td>restrict observation to those obseved in all phenotypes (true) or at least in one phenotype (false)? (default: False)</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd></dd>
<dd>
phenotypes phenotype values </dd>
<dd>
sample_idx_intersect index of individuals in phenotypes after filtering missing values </dd></dl>

</div>
</div>
<a class="anchor" id="a5251a463471ebf0e1d251fdfc1eec6f7"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def limix.modules.data.QTLData.getPos </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>idx_start</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>idx_end</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_start</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_end</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>chrom</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>windowsize</em> = <code>0.0</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>get the positions of the genotypes </p>
<dl class="section return"><dt>Returns</dt><dd></dd>
<dd>
chromosome </dd>
<dd>
position </dd>
<dd>
cumulative_position </dd></dl>

</div>
</div>
<a class="anchor" id="abd07aa418ffd48a352b0661748ef63df"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def limix.modules.data.QTLData.range_query_geno </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>idx_start</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>idx_end</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>chrom</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_start</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_end</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>windowsize</em> = <code>0</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>return an index for a range query on the genotypes </p>

</div>
</div>
<a class="anchor" id="ac5157b8a4ae4a80e933a07c78657cf82"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def limix.modules.data.QTLData.range_query_geno_local </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>idx_start</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>idx_end</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>chrom</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_start</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_end</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>windowsize</em> = <code>0</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>return an index for a range query on the genotypes </p>

</div>
</div>
<a class="anchor" id="af193e0fd43b38bf0c4ac40042c431753"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def limix.modules.data.QTLData.subsample </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>rows</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>cols_pheno</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>cols_geno</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>idx_start</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>idx_end</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_start</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pos_end</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>chrom</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>windowsize</em> = <code>0.0</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>sample a particular set of individuals (rows) or phenotypes (cols_pheno) or genotypes (cols_geno) </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">rows</td><td>indices for a set of individuals </td></tr>
    <tr><td class="paramname">cols_pheno</td><td>indices for a set of phenotypes </td></tr>
    <tr><td class="paramname">cols_geno</td><td>indices for a set of SNPs</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>QTLdata object holding the specified subset of the data </dd></dl>

</div>
</div>
<a class="anchor" id="a52831f8f693a82c852ef68bb8b709528"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def limix.modules.data.QTLData.subsample_phenotypes </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>phenotype_IDs</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>phenotype_query</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>center</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>intersection</em> = <code>False</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>load Phenotypes </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">idx_start</td><td>phenotype indices to load (start individual index) </td></tr>
    <tr><td class="paramname">idx_end</td><td>phenotype indices to load (end individual index) </td></tr>
    <tr><td class="paramname">phenotype_IDs</td><td>names of phenotypes to load </td></tr>
    <tr><td class="paramname">center</td><td>imputation of missing values (default: True) </td></tr>
    <tr><td class="paramname">intersection</td><td>restrict observation to those obseved in all phenotypes (true) or at least in one phenotype (false)? (default: False)</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd></dd>
<dd>
phenotypes phenotype values </dd>
<dd>
sample_idx_intersect index of individuals in phenotypes after filtering missing values </dd></dl>

</div>
</div>
<h2 class="groupheader">Member Data Documentation</h2>
<a class="anchor" id="a44790730592493cfead5620fa01fd2fe"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">limix.modules.data.QTLData.geno_ID</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a89f88a15267238a5891760f65dec36ba"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">limix.modules.data.QTLData.geno_pos</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a9261d716439ec35ab451cf6c6844974b"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">limix.modules.data.QTLData.geno_reader</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a8ec02bd88a4802018e4f50c5ae93f0bc"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">limix.modules.data.QTLData.geno_snp_idx</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a699163ea6b952f2d30e08f270a2383da"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">limix.modules.data.QTLData.num_snps</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a6244c75260ec6d8462f412eab4602c1a"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">limix.modules.data.QTLData.pheno_reader</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a921a453d6b5484595508867ea848459d"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">limix.modules.data.QTLData.phenotype_ID</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="aa205ab51d0684b45bdb25f476afd0256"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">limix.modules.data.QTLData.sample_ID</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a6882068a6f98048ae7ba8ee132b4523d"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">limix.modules.data.QTLData.sample_idx</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<hr/>The documentation for this class was generated from the following file:<ul>
<li>/Users/stegle/research/users/stegle/limix/src/interfaces/python/limix/modules/<a class="el" href="data_8py.html">data.py</a></li>
</ul>
</div><!-- contents -->
<hr>
</body>
</html>
